### DISCUSSION TITLE
How to set the upstream host dynamically

### DISCUSSION THREAD

**Discussion by KamenRiderKuuga** [apache/apisix#9487] [Q&A]
I'm trying to set the upstream host dynamically and unknown in advance like this:

remote_uri: `/api/data-manage/...`
then set the host to: `data-manage`

I've tried the `proxy-rewrite` and the custom balancer, but still can't get this to work.

Is there any suggestion?

**Comment by xBoo** [comment]
I have similar needs to you. My requirement is to parse the tenant ID in the token in the request, then obtain the cluster address corresponding to the tenant from redis based on the tenant ID, and then dynamically route to different upstreams. The following is my part Implementation, for your reference, I hope it can inspire you:

```
function set_upstream_by_reids(ctx, conf, redis_key)

    local red, err2 = redis.new(conf)
    if not red then
        return false, err2
    end

    local cluster, get_err = red:get(redis_key)
    if get_err then
        return false, get_err

    elseif not cluster or cluster == nil or (type(cluster) == "userdata" and tostring(cluster) == "userdata: NULL") then
        core.response.set_header("x-cluster", conf.default_upstream_id)
        ctx.upstream_id = conf.default_upstream_id

        return true
    else
        core.response.set_header("x-cluster", cluster)
        ctx.upstream_id = cluster
        return true
    end
end
```

**Comment by PasanBhanu** [comment]
Hi @KamenRiderKuuga & @xBoo,

I had the same concern and below is how I fixed the issue.

I added proxy-rewrite plugin to capture the parameter from URL and add it to a custom header.

```
"proxy-rewrite": {
      "_meta": {
        "disable": false
      },
      "headers": {
        "add": {
          "custom-header": "$1"
        }
      },
      "regex_uri": [
        "/api/([^/]+)(/.*)$",
        "/api/$1$2"
      ]
    }
```

Then I wrote a custom lua plugin to override the host. This plugin will take the `custom-header` and update the `host` header. Proxy-rewrite plugin will not allow you to use variables in `host` header. That is why I had to write the custom header.

```
local core = require("apisix.core")
local plugin_name = "header-override"

local schema = {
    type = "object",
    properties = {},
}

local _M = {
    version = 0.1,
    priority = 0,
    name = plugin_name,
    schema = schema,
}

function _M.check_schema(conf)
    return core.schema.check(schema, conf)
end

function _M.before_proxy(conf, ctx)
    local headers = ngx.req.get_headers()
    local host = headers["custom-header"]

    if host then
        ctx.var["upstream_host"] = host
        core.log.info("Host header overridden to: ", host)
    else
        core.log.info("custom-header header not found, Host header remains unchanged")
    end
end

return _M
```

Once enabled, this plugin will override the host from `custom-header`.

I have a repository which I used for knative networking, which has the same issue. You can refer it here: [https://github.com/PasanBhanu/apisix-knative-kourier-routing-plugin](https://github.com/PasanBhanu/apisix-knative-kourier-routing-plugin). For the guidance on how to add the custom plugin, you can refer to above repo which has a detailed guideline on how to do that. 

Hope this helps 🫰
